### ISSUE TITLE
[FEA] Abstract away block reduce and block scan from cuIO kernels

### ISSUE THREAD

**Issue by devavret** [rapidsai/cudf#6238]
ORC and parquet writer kernels use a block reduction or a block scan but write out the entire logic inside the kernel.
#### Here's a block reduction:
https://github.com/rapidsai/cudf/blob/57ef76927373d7260b6a0eda781e59a4c563d36e/cpp/src/io/orc/dict_enc.cu#L431-L434

#### Here's a block scan:
https://github.com/rapidsai/cudf/blob/57ef76927373d7260b6a0eda781e59a4c563d36e/cpp/src/io/orc/dict_enc.cu#L266-L284

This code is doing a block scan on dupes_before but the code for this operation is interspersed with other logic.

### Proposed fix
- Either implement a block scan and a block reduction and add it to [block_utils.cuh](https://github.com/rapidsai/cudf/blob/57ef76927373d7260b6a0eda781e59a4c563d36e/cpp/src/io/utilities/block_utils.cuh). It would be helpful if this could be called by just the active threads and not all threads (perhaps with creative use of coalesced threads masks) but not too important as all the places it's called in already ensures `__syncthreads()` can be called.
- Use an off the shelf cub implementation.

**Comment by jrhemstad** [comment]
> Use an off the shelf cub implementation

This is almost certainly the way to go. It's like that the CUB implementations will be faster as well.

**Comment by devavret** [comment]
> > Use an off the shelf cub implementation
> 
> This is almost certainly the way to go. It's likely that the CUB implementations will be faster as well.

I'd like this too but I get tempted by the idea of calling a block reduce by active threads.

**Comment by jrhemstad** [comment]
> I'd like this too but I get tempted by the idea of calling a block reduce by active threads.

Don't all threads contribute to the final reduction result?

**Comment by devavret** [comment]
They do but the last iteration might not. For that purpose, instead of 
```c++
for (size_t i = start; i < end; i += stride) {
  // data[i]
}
```
It currently uses
```c++
for (size_t i = 0; i < size; i += stride) {
  if (i + t < end) data[i + t];
}
```
e.g. here https://github.com/rapidsai/cudf/blob/57ef76927373d7260b6a0eda781e59a4c563d36e/cpp/src/io/orc/dict_enc.cu#L399-L400

**Comment by jrhemstad** [comment]
> They do but the last iteration might not.

Is there useful work for the remaining threads to do after the last iteration where it is actually detrimental to synchronize them?

**Comment by devavret** [comment]
> Is there useful work for the remaining threads to do after the last iteration where it is actually detrimental to synchronize them?

Not for reducing work, this is just for readability of the loop and things inside it. Otherwise most operations would need to have a `if (i + t < end)` before them.

But there are other reasons to allow a `__syncthreads()` inside a loop, like calling another block function like RleEncode  which itself uses `__syncthreads()`. So it's not too important.

Replacing the current code with `cub::BlockReduce` would be a great improvement anyway.

**Comment by kaatish** [comment]
@devavret could using the `valid_items` parameter with cub::BlockReduce ([link](https://nvlabs.github.io/cub/classcub_1_1_block_reduce.html#a33ddffdde07275ab0c4e1bf61b0d9409)) help in handling the last iteration?

**Comment by devavret** [comment]
> @devavret could using the `valid_items` parameter with cub::BlockReduce ([link](https://nvlabs.github.io/cub/classcub_1_1_block_reduce.html#a33ddffdde07275ab0c4e1bf61b0d9409)) help in handling the last iteration?

If you mean that it might obviate the need to set out of bounds elements to 0 then it appears so from the docs. If you mean it might allow us to write our loops in `for (size_t i = start; i < end; i += stride) {` style then perhaps not. cub::BlockReduce is [not syncthreads free](https://github.com/NVlabs/cub/blob/c3cceac115c072fb63df1836ff46d8c60d9eb304/cub/block/specializations/block_reduce_raking.cuh#L182).

**Comment by OlivierNV** [comment]
Keep in mind that a lot of what cub does is abstract away differences between pre-sm_50 and maxwell+ (existence shfl).
Perhaps I missed something, but my early experiments with cub reductions invariably showed extremely high shared mem usage (like 4K smem storage for a block reduction) and much higher register usage that the explicit equivalent butterfly shuffle reduction (this is largely why I did not end up relying on cub in cuio kernels).

**Comment by harrism** [comment]
If that's the case we can file a CUB issue and get it fixed (or fix it ourselves and contribute it). Sharpen our tools.

**Comment by jrhemstad** [comment]
I doubt there is anything to fix. 

CUB has 3 different algorithms you can select for the block reduction: https://nvlabs.github.io/cub/namespacecub.html#add0251c713859b8974806079e498d10a

The pros/cons of each are described in the documentation. Shmem/register usage only matter if they negatively impact occupancy.

**Comment by OlivierNV** [comment]
Just trying to help. All I'm saying is make sure to check such assumptions against reality.

**Comment by devavret** [comment]
> Just trying to help. All I'm saying is make sure to check such assumptions against reality.

How should one check these? What's your tool of choice? NSight compute￼/ cuobjdump? I'm asking because I haven't checked shared mem usage or register usage of a￼ kernel before.

**Comment by jrhemstad** [comment]
> How should one check these? What's your tool of choice? NSight compute￼/ cuobjdump? I'm asking because I haven't checked shared mem usage or register usage of a￼ kernel before.

`cubobjdump --dump-resource-usage` is the easiest way to see the raw resource usage of a kernel. Nsight Compute can then be used to see if any change in resource usage affected occupancy.

**Comment by OlivierNV** [comment]
Personally, I always use the ptxas -v option at build time during development (can't imagine doing any serious cuda work without it). I also consider any non-zero lmem usage a bug with very rare exceptions.

**Comment by vuule** [comment]
Somehow this issue hasn't closed when #7278 was merged, closing manually.
